auth_server/src/main.rs

```diff
@@ -39,23 +48,64 @@ async fn main() -> Result<()> {
 
 async fn handle_incoming_connection(mut stream: TcpStream, auth_database: std::sync::Arc<AuthDatabase>) -> Result<()> {
     info!("incoming on address {}", stream.local_addr()?.to_string());
-    let mut logindata = auth::LoginNumbers::default();
+    let mut challenge_proof: Option<ServerChallengeProof> = None;
+    let mut log_on_proof: Option<ServerLogOnProof> = None;
 
+    //TODO(wmxd): keep some state that so the packets cant come out of order
```

Comment:
The protocol uses TCP. Can packets even arrive out of order?

Reply:
I mean that if you have rogue client that send logon proof before challenge, to that actual packets from network will be out of order.

Reply:
It will be some kind of state machine later like:
challenge -> logon -> realm list
reconnect challenge -> (if we got the session_key for username) reconnect logon -> realm list